akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala

```diff
@@ -132,6 +134,152 @@ import scala.util.control.NonFatal
+          subSource.complete()
+        } else {
+          materializeFlow()
+        }
+      }
+
+      override def onUpstreamFailure(ex: Throwable): Unit = {
+        if (null != subSource) {
+          subSource.fail(ex)
+        } else {
+          //flow won't be materialized, so we have to complete the future with a failure indicating this
+          matPromise.failure(new NeverMaterializedException(ex))
+          super.onUpstreamFailure(ex)
+        }
+      }
+
+      override def onPull(): Unit = {
+        if (null != subSink) {
+          //delegate to subSink
+          subSink.pull()
+        } else if (accumulated.size < n) {
+          pull(in)
+        } else if (accumulated.size == n) { //corner case for n = 0, can be handled in FlowOps
+          materializeFlow()
+        }
+      }
+
+      override def onDownstreamFinish(cause: Throwable): Unit = {
+        if (null != subSink) {
+          subSink.cancel(cause)
+        } else {
+          materializeFlow()
+        }
+      }
+
+      def materializeFlow(): Unit = {
+        val prefix = accumulated.reverse
+        accumulated = Nil
+        subSource = new SubSourceOutlet[In]("subSource")
+        subSource.setHandler {
+          new OutHandler {
+            override def onPull(): Unit = {
+              if (!isClosed(in) && !hasBeenPulled(in)) {
+                pull(in)
+              }
+            }
+
+            override def onDownstreamFinish(cause: Throwable): Unit = {
+              if (!isClosed(in)) {
+                cancel(in, cause)
+              }
+              //super.onDownstreamFinish(cause)
+            }
+          }
+        }
+        subSink = new SubSinkInlet[Out]("subSink")
+        subSink.setHandler {
+          new InHandler {
+            override def onPush(): Unit = {
+              push(out, subSink.grab())
+            }
+
+            override def onUpstreamFinish(): Unit = {
+              complete(out)
+            }
+
+            override def onUpstreamFailure(ex: Throwable): Unit = {
+              fail(out, ex)
+            }
+          }
+        }
+        val matValueTry = Try {
+          val flow = f(prefix)
+          val runnableGraph = Source.fromGraph(subSource.source).viaMat(flow)(Keep.right).to(subSink.sink)
+          interpreter.subFusingMaterializer.materialize(runnableGraph)
+        }
+
+        matPromise.complete {
+          matValueTry.recoverWith {
+            case ex => scala.util.Failure(new NeverMaterializedException(ex))
+          }
+        }
+
+        matValueTry.foreach { _ =>
+          //in case we've materialized due to upstream completion
+          if (isClosed(in)) {
+            subSource.complete()
+          }
+          //in case we've been pulled by downstream
+          if (isAvailable(out)) {
+            subSink.pull()
+          }
+        }
+
+        matValueTry.failed.foreach { ex =>
+          subSource = null
+          subSink = null
+          failStage(ex)
+        }
```

Comment:
I still think this whole section would be cleaner with a try-catch doing the same stuff than the `Try` wrapping, it would also save 4 lambda allocs.